File: 16-3sum-closest/16-3sum-closest.java
class Solution {
    public int threeSumClosest(int[] nums, int target) {
        int low = 0, high = 0, res = target >= 0 ? Integer.MAX_VALUE: Integer.MIN_VALUE, sum = 0;
        Arrays.sort(nums);
        for(int i = 0; i < nums.length - 2; i ++){
            low = i + 1; high = nums.length - 1;
            while(low < high){
                sum = nums[i] + nums[low] + nums[high];
                if(sum == target) return target;
                else{
                    res = Math.abs(sum - target) < Math.abs(res - target) ? sum: res;
                    if(sum < target) low++;
                    else high--;
                }
            }
        }
        return res;
    }
}

File: 16-3sum-closest/16-3sum-closest.cpp
class Solution {
public:
    int threeSumClosest(vector<int>& nums, int target) {
        int res = nums[0]+nums[1]+nums[2];
        int min = abs(nums[0]+nums[1]+nums[2]-target);
        if(min==0)
            return target;
            
        sort(nums.begin(), nums.end());
        
        for(int i=0; i<nums.size()-2; i++){
            int front = i+1;
            int back = nums.size()-1;
            
            while(front<back){
                int sum = nums[i]+nums[front]+nums[back];
                if(sum==target)
                    return target;
                
                if(abs(sum-target)<min){
                    min = abs(sum-target);
                    res = sum;
                    int minfront = front;
                    int minback = back;
                    // while(front<back && nums[front]==nums[minfront])
                    //     front++;
                    // while(front<back && nums[back]==nums[minback])
                    //     back--;
                }  
                else{
                    if(sum<target){
                        front++;
                    }
                    else if(sum>target){
                        back--;
                    }                                 
                }

            }
            
            while(i+1<nums.size() && nums[i+1]==nums[i])
                i++;
            
        }
        
        return res;
    }
};
